folding-schemes/src/folding/nova/nifs/mova.rs

```diff
@@ -0,0 +1,415 @@
+        })
+    }
+}
+
+#[derive(Debug, Clone, Eq, PartialEq, CanonicalSerialize, CanonicalDeserialize)]
+pub struct Proof<C: CurveGroup> {
+    pub h_proof: PointVsLineProof<C>,
+    pub mleE1_prime: C::ScalarField,
+    pub mleE2_prime: C::ScalarField,
+    pub mleT: C::ScalarField,
+}
+
+/// Implements the Non-Interactive Folding Scheme described in section 4 of
+/// [Mova](https://eprint.iacr.org/2024/1220.pdf).
+/// `H` specifies whether the NIFS will use a blinding factor
+pub struct NIFS<
+    C: CurveGroup,
+    CS: CommitmentScheme<C, H>,
+    T: Transcript<C::ScalarField>,
+    const H: bool = false,
+> {
+    _c: PhantomData<C>,
+    _cp: PhantomData<CS>,
+    _ct: PhantomData<T>,
+}
+
+impl<C: CurveGroup, CS: CommitmentScheme<C, H>, T: Transcript<C::ScalarField>, const H: bool>
+    NIFSTrait<C, CS, T, H> for NIFS<C, CS, T, H>
+where
+    <C as Group>::ScalarField: Absorb,
+    <C as CurveGroup>::BaseField: PrimeField,
+{
+    type CommittedInstance = CommittedInstance<C>;
+    type Witness = Witness<C>;
+    type ProverAux = Vec<C::ScalarField>; // T in Mova's notation
+    type Proof = Proof<C>;
+
+    fn new_witness(w: Vec<C::ScalarField>, e_len: usize, rng: impl RngCore) -> Self::Witness {
+        Witness::new::<H>(w, e_len, rng)
+    }
+
+    fn new_instance(
+        mut rng: impl RngCore,
+        params: &CS::ProverParams,
+        W: &Self::Witness,
+        x: Vec<C::ScalarField>,
+        aux: Vec<C::ScalarField>, // = r_E
+    ) -> Result<Self::CommittedInstance, Error> {
+        let mut rE = aux.clone();
+        if is_zero_vec(&rE) {
+            // means that we're in a fresh instance, so generate random value
+            rE = (0..log2(W.E.len()))
+                .map(|_| C::ScalarField::rand(&mut rng))
+                .collect();
+        }
+
+        W.commit::<CS, H>(params, x, rE)
+    }
+
+    // Protocol 7 - point 3 (16)
+    fn fold_witness(
+        a: C::ScalarField,
+        W_i: &Witness<C>,
+        w_i: &Witness<C>,
+        aux: &Vec<C::ScalarField>, // T in Mova's notation
+    ) -> Result<Witness<C>, Error> {
+        let a2 = a * a;
+        let E: Vec<C::ScalarField> = vec_add(
+            &vec_add(&W_i.E, &vec_scalar_mul(aux, &a))?,
+            &vec_scalar_mul(&w_i.E, &a2),
+        )?;
+        let W: Vec<C::ScalarField> = W_i
+            .W
+            .iter()
+            .zip(&w_i.W)
+            .map(|(i1, i2)| *i1 + (a * i2))
+            .collect();
+
+        let rW = W_i.rW + a * w_i.rW;
+        Ok(Witness::<C> { E, W, rW })
+    }
+
+    /// [Mova](https://eprint.iacr.org/2024/1220.pdf)'s section 4. Protocol 8
+    /// Returns a proof for the pt-vs-line operations along with the folded committed instance
+    /// instances and witness
+    #[allow(clippy::type_complexity)]
+    fn prove(
+        _cs_prover_params: &CS::ProverParams, // not used in Mova since we don't commit to T
+        r1cs: &R1CS<C::ScalarField>,
+        transcript: &mut T,
+        pp_hash: C::ScalarField,
```

Comment:
This is used for challenges that come from transcript/hashing, the pp_hash itself is a hash of several public values of the parameters (such as the circuit r1cs, the commitment schemes params, the poseidon config, etc) so that the challenges are dependent on the specific setup (and the same challenge is not generated between 2 different curves instantiations for example). In Nova & Mova papers I think that they appear denoted as `pp` (for 'public parameters').